### ISSUE TITLE
Generating URL with page parameter fails with new endpoint routing (regression from 2.1)

### ISSUE THREAD

**Issue by poke** [aspnet/AspNetCore#5055]
This appears to be a regression due to the new endpoint routing system in 2.2. As per aspnet/Mvc#7876, using parameter names `action` and `page` is somewhat difficult, however using a `page` parameter inside an MVC action or using an `action` parameter inside of a Razor page does work fine. At least it did.

With endpoint routing activated (through the MVC compatibility switch `CompatibilityVersion.Version_2_2`), generating an action link that has a `page` parameter will now fail **iff** Razor pages exist in the same project.

#### Without Razor pages
1. Create new 2.2 application from standard MVC template.
2. Add an action to the `HomeController`:
    ```c#
    public IActionResult Test(int page)
    {
        return Json(new
        {
            Url = Url.Action(nameof(Test), new { page = 123 }),
            Page = page,
        });
    }
    ```
3. Launch the application and open `http://localhost:5000/Home/Test?page=4`
4. Observe the result which shows a generated URL `/Home/Test?page=123`.

#### Add Razor pages
5. Now, create a Razor page at `Pages/Foo.cshtml` with the contents `@page`.
6. Launch the application and open the previous URL again.
7. Observe that the generated URL is now `null`.

#### Disable endpoint routing
8. Disable endpoint routing by changing the `AddMvc()` call in `Startup` to set the compatibility level switch back to 2.1:
    ```c#
    services.AddMvc().SetCompatibilityVersion(CompatibilityVersion.Version_2_1);
    ```
9. Launch the application and open the previous URL again.
10. Observe that the URL is generated properly again.

---

As you can see, endpoint routing breaks the URL generation here although the target URL exists just fine and although it also can be used properly the whole time (note that the JSON response always includes the passed `page` value). The mere presence of Razor pages should not make the URL generation treat any `page` parameter in a special way here.

**Comment by nickalbrecht** [comment]
I ran into this yesterday/today. Thanks for taking the time to flesh out the details and submit the bug @poke. My short term solution was to rename my action parameters to something like `pageNum` instead of `page`, But I'd prefer to see this get fixed so it wasn't so aggressive with regards to preventing the use of `page` as an action parameter when trying to generate URLs.

**Comment by WistfulWolf** [comment]
Just encountered this issue in .NET Core 2.1. UrlHelper.Link when adding a parameter with name "page" completely ignores it. Tested on clean MVC project with an API controller.

Api Controller class:
```
[ApiController, Route("api/[controller]")]
public class TestController : Controller
{
    [HttpGet(Name = "Test")]
    public void Test()
    {
    }
}
```
Page class:
```
public class AboutModel : PageModel
{
    public string Message { get; set; }

    public void OnGet()
    {
        string url = Url.Link("T", new { page = 1 });
    }
}
```

The result is **url = "http://localhost/api/Test"**
Expected result is **url = "http://localhost/api/Test?page=1"**

**Comment by mkArtakMSFT** [comment]
Thanks for contacting us, @poke.
@JamesNK, can you please look into this? Thanks!

**Comment by mkArtakMSFT** [comment]
Ping @JamesNK!

**Comment by JamesNK** [comment]
I can reproduce this with endpoint routing. The issue comes from the new link generator requiring that `page` value be null because of the null `page` required value from MVC. A non-null value and the decision tree can't resolve the endpoint.

I tried a couple of minor changes to `LinkGenerationDecisionTree`, and I could get the link to successfully generate, but not without breaking at least some other routes. I'm not familiar with the inner workings of the link generator. @rynowak do you have some time to pair program this with me?

Test plus some broken attempts: https://github.com/aspnet/AspNetCore/commit/b493845bf104a422a28a76517659ee87a60fa4d1

**Comment by JamesNK** [comment]
Ryan and I looked into this issue. We don't think it is something that can be easily patched in 2.2. A fix will come in 3.0

If you have this problem then you should be able to fix it in 2.2 by disabling endpoint routing in Startup.cs where `AddMvc` is called:

`services.AddMvc(options.EnableEndpointRouting = false)`

**Comment by rynowak** [comment]
This primary case reported here has been fixed by: https://github.com/aspnet/AspNetCore/pull/7441 for preview 3.

Note that there may still be a little bit of wierdness - in particular using `page` as a route parameter along with other route parameters may not preserve the ambient values.

This issue is difficult to solve because *fundamentally* the routing system uses the presence or absence of these values in its decisions about what link to try and generate. I'm going to revisit the design of this and see what else I can come up with since this is a common problem. The changes in preview3 will hopefully make this a little bit better.
